Mathlib/Algebra/Group/Action/Equidecomp.lean

```diff
@@ -0,0 +1,233 @@
+`PartialEquiv X X` with source `A` and target `B`.
+
+## Implementation Notes
+
+* Equidecompositions are implemented as elements of `PartialEquiv X X` together with a
+  `Finset` of elements of the acting group and a proof that every point in the source is moved
+  by an element in the finset.
+
+* The requirement that `G` be a group is relaxed where possible.
+
+* We introduce a non-standard predicate, `IsDecompOn`, to state that a function satisfies the main
+  combinatorial property of equidecompositions, even if it is not injective or surjective.
+
+## TODO
+
+* Prove that if two sets embeddidecompose into eachother, they are equidecomposable
+  (Schroeder-Bernstein type theorem)
+* Define embeddidecomposability as a Preorder on sets and prove that its induced equivalence
+  relation is equidecomposability.
+* Prove the definition of equidecomposition used here is equivalent to the more familiar one
+  using partitions.
+
+-/
+
+variable {X G : Type*} {A B C: Set X}
+
+open Function Set Pointwise PartialEquiv
+
+namespace Equidecomp
+
+section SMul
+
+variable [SMul G X]
+
+/-- Let `G` act on a space `X` and `A : Set X`. We say `f : X → X` is a decomposition on `A`
+as witnessed by some `S : Finset G` if for all `a ∈ A`, the value `f a` can be obtained
+by applying some element of `S` to `a` instead.
+
+More familiarly, the restriction of `f` to `A` is the result of partitioning `A` into finitely many
+pieces, then applying a single element of `G` to each piece. -/
+def IsDecompOn (f : X → X) (A : Set X) (S : Finset G) : Prop := ∀ a ∈ A, ∃ g ∈ S, f a = g • a
+
+variable (X G)
+
+/-- Let `G` act on a space `X`. An `Equidecomposition` with respect to `X` and `G` is a partial
+bijection `f : PartialEquiv X X` with the property that for some set `elements : Finset G`,
+(which we record), for each `a ∈ f.source`, `f a` can be obtained by applying some `g ∈ elements`
+instead. We call `f` an equidecomposition of `f.source` with `f.target`.
+
+More familiarly, `f` is the result of partitioning `f.source` into finitely many pieces,
+then applying a single element of `G` to each to get a partition of `f.target`.
+-/
+structure _root_.Equidecomp extends PartialEquiv X X where
+  decomp' : ∃ S : Finset G, IsDecompOn toFun source S
+
+variable {X G}
+
+/-- Note that `Equidecomp X G` is not `FunLike`. -/
+instance : CoeFun (Equidecomp X G) fun _ => X → X := ⟨fun f => f.toFun⟩
+
+/-- A finite set of group elements witnessing that `f` is an equidecomposition. -/
+noncomputable
+def elements (f : Equidecomp X G) := f.decomp'.choose
+
+theorem decomp (f : Equidecomp X G) : IsDecompOn f f.source f.elements := f.decomp'.choose_spec
+
+@[simp]
+theorem map_source {f : Equidecomp X G} {x : X} (h : x ∈ f.source) :
+    f x ∈ f.target := f.toPartialEquiv.map_source h
+
+theorem toPartialEquiv_injective : Injective <| toPartialEquiv (X := X) (G := G) := by
+  intro f f' _
+  let ⟨_, _, _⟩ := f
+  congr
```

Comment:
Potential golf, but I'm unsure if this fits with the style guidelines (@YaelDillies?)

```suggestion
theorem toPartialEquiv_injective : Injective <| toPartialEquiv (X := X) (G := G) := by
  intro ⟨_, _, _⟩ _ _
  congr
```

Reply:
Definitely fits, yes!